edgedb-protocol/src/serialization/decode/raw_composite.rs

```diff
@@ -0,0 +1,197 @@
+use crate::queryable::Queryable;
+use crate::errors::{self, DecodeError};
+use snafu::ensure;
+use self::inner::DecodeCompositeInner;
+
+
+pub struct DecodeTupleLike<'t> {
+    inner:DecodeCompositeInner<'t>
+}
+
+impl<'t> DecodeTupleLike<'t> {
+    fn new(buf:&'t [u8]) -> Result<Self, DecodeError> {
+        let inner = DecodeCompositeInner::read_tuple_like_header(buf)?;
+        Ok(DecodeTupleLike{inner})
+    }
+
+    pub fn new_object(buf:&'t [u8], expected_count:usize) -> Result<Self, DecodeError> {
+        let elements = Self::new(buf)?;
+        ensure!(elements.inner.count() == expected_count, errors::ObjectSizeMismatch);
+        Ok(elements)
+    }
+
+    pub fn new_tuple(buf:&'t [u8], expected_count:usize) -> Result<Self, DecodeError> {
+        let elements = Self::new(buf)?;
+        ensure!(elements.inner.count() == expected_count, errors::TupleSizeMismatch);
+        Ok(elements)
+    }
+
+    pub fn read(&mut self) -> Result<Option<&[u8]>, DecodeError> {
+        self.inner.read_object_element()
+    }    
+
+    pub fn decode_element<T:Queryable>(&mut self) -> Result<T, DecodeError> {
+        let element = self.read()?;
+        T::decode_optional(element)
+    }
+
+    pub fn skip_element(&mut self) -> Result<(), DecodeError> {
+        self.read()?;
+        Ok(())
+    }
+}
+
+pub struct DecodeInputTuple<'t> {
+    inner:DecodeCompositeInner<'t>
+}
+
+impl<'t> DecodeInputTuple<'t> {
+    fn new(buf:&'t [u8]) -> Result<Self, DecodeError> {    
+        let inner = DecodeCompositeInner::read_tuple_like_header(buf)?;
+        Ok(DecodeInputTuple{inner})
+    }
+
+    pub fn with_count(buf:&'t [u8], expected_count:usize) -> Result<Self, DecodeError> {
+        let elements = Self::new(buf)?;
+        ensure!(elements.inner.count() == expected_count, errors::TupleSizeMismatch);
+        Ok(elements)
+    }
+
+    pub fn read(&mut self) -> Result<&[u8], DecodeError> {
+        self.inner.read_array_like_element()
+    }    
+}
+
+pub struct DecodeArrayLike<'t> {
+    inner:DecodeCompositeInner<'t>
+}
+
+impl<'t> DecodeArrayLike<'t> {
+    pub fn new_array(buf:&'t [u8]) -> Result<Self, DecodeError> {
+        let inner = DecodeCompositeInner::read_array_like_header(buf, || errors::InvalidArrayShape.fail::<()>().err().unwrap())?;

```

Comment:
nit: We try to maintain 79 characters per line limit to avoid using wordwrap in editors and make it possible to fit a few editing panes with code on one monitor.